dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,263 @@
+﻿#include "pch.h"
+
+#include "PushNotificationManager.h"
+#include "Microsoft.Windows.PushNotifications.PushNotificationManager.g.cpp"
+
+#include "PushNotificationCreateChannelResult.h"
+
+#include <winrt/Windows.ApplicationModel.background.h>
+#include <winrt\windows.networking.pushnotifications.h>
+#include "PushNotificationBackgroundTask.h"
+
+#include <winerror.h>
+#include <algorithm>
+#include "PushNotificationChannel.h"
+#include "externs.h"
+
+constexpr PCWSTR backgroundTaskName = L"PushBackgroundTaskName";
+
+namespace winrt
+{
+    using namespace Windows::ApplicationModel::Background;
+    using namespace Windows::Networking::PushNotifications;
+    using namespace Windows::Foundation;
+}
+
+wil::critical_section g_lock;
+
+namespace winrt::Microsoft::Windows::PushNotifications::implementation
+{
+    inline constexpr std::uint32_t c_maxBackoffSeconds{ 960 };
+    inline constexpr std::uint32_t c_minBackoffSeconds{ 30 };
+
+    const HRESULT WNP_E_NOT_CONNECTED = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403E8L);
+    const HRESULT WNP_E_RECONNECTING = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403E9L);
+    const HRESULT WNP_E_BIND_USER_BUSY = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403FEL);
+
+    bool PushNotificationManager::isChannelRequestRetryable(const hresult& hr)
+    {
+        switch (hr)
+        {
+        case HRESULT_FROM_WIN32(ERROR_TIMEOUT):
+        case WNP_E_NOT_CONNECTED:
+        case WPN_E_OUTSTANDING_CHANNEL_REQUEST:
+        case WNP_E_RECONNECTING:
+        case WNP_E_BIND_USER_BUSY:
+        case RPC_S_SERVER_UNAVAILABLE:
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    inline bool IsPackagedProcess()
+    {
+        wchar_t packageFullName[PACKAGE_FULL_NAME_MAX_LENGTH + 1] = {};
+
+        UINT32 packageFullNameLength = static_cast<UINT32>(ARRAYSIZE(packageFullName));
+
+        const auto packagedProcessHResult = ::GetCurrentPackageFullName(&packageFullNameLength, packageFullName);
+
+        if (packagedProcessHResult == APPMODEL_ERROR_NO_PACKAGE)
+        {
+            return false;
+        }
+
+        THROW_IF_WIN32_ERROR(packagedProcessHResult);
+
+        return true;
+    }
+
+    winrt::IAsyncOperationWithProgress<winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelResult, winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelStatus> PushNotificationManager::CreateChannelAsync(winrt::guid remoteId)
+    {
+        static bool s_remoteIdInProgress;
```

Comment:
So channel creations - even for different channel IDs - cannot run at the same time?  An app has to synchronize each of its CreateChannel requests with each other?

Reply:
Today, a single packaged app can request only a single channel with a remoteId. So always a one to one mapping. We don't support one to many. The intention here is to prevent subsequent channel calls from the same app if one is already in progress.

Reply:
What about asking for the same channel twice? Is that allowed? (Does it return the same channel back?)

It would not be unreasonable for different parts of a single app to want to listen to the notification channel for different notifications. (The game matchmaker wants to listen for matchmaking notifications. The high score manager wants to know if the user's high score has been beaten, etc.) It would be nice if they could each get the channel and do their own notification filtering. (Especially if the "high score manager" is some third party library that has no way of coordinating with the app's matchmaker component.)

Reply:
It depends: If the first channel request is already in progress but hasn't completed yet, the first one would return a Completion event with a specific Throttled error code. 

If the first channel request has completed, the second one would progress and return a new channel from the platform.

Note: There are certain caching intracies built into the OS if the same channel is request within 24hrs of the prior one. We would simply return the cached result in our DB for the < 24hr case. For > 24hr case, we would send a request over the wire to our service to get a new channel.